services/version.js

```diff
@@ -93,10 +91,29 @@ function latest(versions, { pre = false } = {}) {
   } catch (e) {
     version = latestDottedVersion(versions)
   }
-  if (version === undefined || version === null) {
-    origVersions = origVersions.sort()
-    version = origVersions[origVersions.length - 1]
+  return version
+}
+
+// Given a list of versions (as strings), return the latest version.
+// Return undefined if no version could be found.
+function latest(versions, { pre = false } = {}) {
+  let version = ''
+
+  version = latestMaybeSemVer(versions, pre)
+
+  if (version == null && !pre) {
+    version = latestMaybeSemVer(versions, true)
   }
+
+  // if we've still got nothing,
+  // fall back to a case-insensitive string comparison
+  if (version == null) {
+    versions = versions.sort((a, b) =>
+      a.toLowerCase().localeCompare(b.toLowerCase())
+    )
+    version = versions[versions.length - 1]
+  }
+
```

Comment:
I like this flow from a readability perspective, but seems like it could add some additional iterations over the input `versions` in the fallback cases. 

I don't have a good handle on how frequently we have to drop to the fallback case, nor in those cases how mange pre-release versions there typically are (I'd guess/hope not significant amount).

I'd guess it's not likely to have a big impact, but would be good IMO if there was a clean way to derive the full set semver-compatible versions and, if need be, the single numbered versions one time and reuse them in the fallback case.

Reply:
Rather than using regex, we could check `semver.valid()` on every version, although implicitly that's already happening in `rcompare()`: if it fails, we throw.

Reply:
I'll preface with the usual disclaimer about code written inline within the GitHub Web UI, but I guess I'm wondering if we could do something like:

```js
function latest(versions, { pre = false } = {}) {
  let version = ''
  const origVersions = versions
  // return all results that are likely semver compatible versions
  versions = origVersions.filter(version => /\d+\.\d+/.test(version))
  // If no semver versions then look for single numbered versions
  if (!versions.length) {
    versions = origVersions.filter(version => /\d+/.test(version))
  }

  version = latestMaybeSemVer(versions, pre)

  if (version == null && !pre) {
    version = latestMaybeSemVer(versions, true)
  }

  // if we've still got nothing,
  // fall back to a case-insensitive string comparison
  if (version == null) {
    versions = versions.sort((a, b) =>
      a.toLowerCase().localeCompare(b.toLowerCase())
    )
    version = versions[versions.length - 1]
  }

  return version
}

function latestMaybeSemVer(versions, pre) {
  let version = ''

  if (!pre) {
    // remove pre-releases from array
    versions = versions.filter(version => !/\d+-\w+/.test(version))
  }
  try {
    // coerce to string then lowercase otherwise alpha > RC
    version = versions.sort((a, b) =>
      semver.rcompare(
        `${a}`.toLowerCase(),
        `${b}`.toLowerCase(),
        /* loose */ true
      )
    )[0]
  } catch (e) {
    version = latestDottedVersion(versions)
  }
  return version
}
```

Reply:
I'll also readily admit this could be an unnecessary/premature optimization consideration. I just wanted to point out that it seemed like we'd be adding some extra loops in the fallback case, and I'm thinking we should be able to solve the sorting issue without necessarily having to do so.